File: src/main/java/uy/kohesive/elasticsearch/dataimport/udf/Udfs.java
package uy.kohesive.elasticsearch.dataimport.udf;

import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.api.java.UDF1;
import org.apache.spark.sql.api.java.UDF2;
import org.apache.spark.sql.types.DataTypes;

public class Udfs {

    public static <T extends String, RT extends String> void registerStringToStringUdf(final SparkSession spark, final String name, final UDF1<T, RT> f) {
        spark.udf().register(name, f, DataTypes.StringType);
    }

    public static <T1 extends Object, T2 extends Object, RT extends Object> void registerAnyAnyToTimestampUdf(final SparkSession spark, final String name, final UDF2<T1, T2, RT> f) {
        spark.udf().register(name, f, DataTypes.TimestampType);
    }

    public static <T1 extends String, T2 extends String, RT extends String> void registerStringStringToStringUdf(final SparkSession spark, final String name, final UDF2<T1, T2, RT> f) {
        spark.udf().register(name, f, DataTypes.StringType);
    }


}

File: src/main/kotlin/uy/kohesive/elasticsearch/dataimport/Udfs.kt
package uy.kohesive.elasticsearch.dataimport

import org.apache.spark.sql.SparkSession
import org.jsoup.Jsoup
import org.jsoup.parser.Parser
import org.jsoup.safety.Whitelist
import uy.kohesive.elasticsearch.dataimport.udf.Udfs
import java.sql.Date
import java.sql.Timestamp

object DataImportHandlerUdfs {
    fun registerSparkUdfs(spark: SparkSession) {
        Udfs.registerStringToStringUdf(spark, "fluffly", fluffly)
        Udfs.registerStringToStringUdf(spark, "stripHtml", stripHtmlCompletely)
        Udfs.registerStringToStringUdf(spark, "normalizeQuotes", normalizeQuotes)
        Udfs.registerStringToStringUdf(spark, "unescapeHtmlEntites", unescapeHtmlEntities)
        Udfs.registerAnyAnyToTimestampUdf(spark, "combineDateTime", combineDateTime)
    }

    @JvmStatic val whiteListMap = mapOf(
            "none" to Whitelist.none(),
            "basic" to Whitelist.basic(),
            "basicwithimages" to Whitelist.basicWithImages(),
            "relaxed" to Whitelist.relaxed(),
            "simpletext" to Whitelist.simpleText(),
            "simple" to Whitelist.simpleText()
    )

    @JvmStatic val fluffly = fun (v: String): String = "fluffly " + v

    @JvmStatic val stripHtmlCompletely = fun (v: String): String {
        return Jsoup.parseBodyFragment(v).text()
    }

    @JvmStatic val normalizeQuotes = fun (v: String): String {
        return v.replace("\\'", "'").replace("''", "\"")
    }

    @JvmStatic val unescapeHtmlEntities = fun (v: String): String {
        return Parser.unescapeEntities(v, false)
    }

    @JvmStatic val combineDateTime = fun (date: Date?, time: Timestamp?): Timestamp? {
        // https://stackoverflow.com/questions/26649530/merge-date-and-time-into-timestamp
        if (date == null || time == null) {
            return null
        }

//        val dd = (date.time / 86400000L * 86400000L) - date.timezoneOffset * 60000
//        val tt = time.time - time.time / 86400000L * 86400000L
//        return Timestamp(dd + tt)

        return Timestamp(date.year, date.month, date.date, time.hours, time.minutes, time.seconds, 0)
    }

}